packages/compiler/core/util.ts

```diff
@@ -322,3 +322,95 @@ class CaseInsensitiveMap<T> extends Map<string, T> {
     return super.delete(key.toUpperCase());
   }
 }
+
+interface OrderedMapKey {
+  key: string;
+}
+
+export class OrderedMap<V> implements Map<string, V> {
+  #keys = new Map<string, OrderedMapKey>();
+  #values = new Map<OrderedMapKey, V>();
+
+  constructor(entries: [string, V][]) {
+    for (const [key, value] of entries) {
+      this.set(key, value);
+    }
+  }
+
+  clear(): void {
+    this.#keys.clear();
+    this.#values.clear();
+  }
+
+  delete(key: string): boolean {
+    const keyItem = this.#keys.get(key);
+    if (keyItem) {
+      this.#keys.delete(key);
+      return this.#values.delete(keyItem);
+    }
+    return false;
+  }
+
+  forEach(callbackfn: (value: V, key: string, map: Map<string, V>) => void, thisArg?: any): void {
+    this.#values.forEach((value, keyItem, map) => {
+      callbackfn(value, keyItem.key, this);
+    }, thisArg);
+  }
+  get(key: string): V | undefined {
+    const keyItem = this.#keys.get(key);
+    return keyItem ? this.#values.get(keyItem) : undefined;
+  }
+
+  has(key: string): boolean {
+    const keyItem = this.#keys.get(key);
+    return keyItem ? this.#values.has(keyItem) : false;
+  }
+
+  set(key: string, value: V): this {
+    let keyItem = this.#keys.get(key);
+    if (keyItem === undefined) {
+      keyItem = { key };
+      this.#keys.set(key, keyItem);
+    }
+
+    this.#values.set(keyItem, value);
+    return this;
+  }
+
+  get size() {
+    return this.#values.size;
+  }
+
+  entries(): IterableIterator<[string, V]> {
```

Comment:
These should probably be implemented as generators so we don't create the intermediate array in memory just to peel off an iterator for its items.

Reply:
generator or iterator? I was under the impression that generators have pretty bad performance

Reply:
Generator, hand-rolling an iterator would require some pretty intense transforms that would definitely be slower. I don't think generators are that slow anymore, but regardless I would say that for a general purpose data structure the allocation of the intermediate array is the worse evil.